File: 416.分割等和子集.java
/*
 * @lc app=leetcode.cn id=416 lang=java
 *
 * [416] 分割等和子集
 */

// @lc code=start
class Solution {
    public boolean canPartition(int[] nums) {
        int n = nums.length;
        int sum = 0;
        int max = Integer.MIN_VALUE;
        for (int num : nums) {
            sum += num;
            if(max<num)max= num;
        }
        int target = sum/2;

        if(sum%2==1||max>target){
            return false;
        }
        if(max==target)return true;
        // dp表示0.。i的数中和为j的有没有
        boolean[][] dp = new boolean[n][target+1];
        dp[0][nums[0]]=true;
        for(int i=1;i<n;i++){
            for(int j=1;j<=target;j++){
                // 对当前j可以选择nums[i]
                if(j-nums[i]>=0){
                    // 分为 选择和不选择两种情况,
                    // 找到i前面有没有数的和加起来再加上nums[i]和为j
                    // 如果dp[i-1][j]可以满足情况
                    dp[i][j] = dp[i-1][j-nums[i]]||dp[i-1][j];
                }
                // 不选择nums[i]
                else{
                    dp[i][j]=dp[i-1][j];
                }
            }
        }
        return dp[n-1][target];
    }
}
// @lc code=end

File: 416.分割等和子集.py
#
# @lc app=leetcode.cn id=416 lang=python3
#
# [416] 分割等和子集
#
from typing import List
# @lc code=start


class Solution:
    def canPartition2(self, nums: List[int]) -> bool:
        n = len(nums)
        if n < 2:
            return False
        sum = 0
        max_num = 0
        for num in nums:
            sum += num
            if max_num < num:
                max_num = num
        target = sum//2

        if sum % 2 or max_num > target:
            return False
        # dp[i][j]表示0...i中和为j的情况
        dp = [[True]+[False]*target for _ in range(n)]
        if target >= nums[0]:
            dp[0][nums[0]] = True
        for i in range(1, n):
            for j in range(1, target+1):
                if j < nums[i]:
                    dp[i][j] = dp[i-1][j]
                else:
                    dp[i][j] = dp[i-1][j-nums[i]] | dp[i-1][j]
        return dp[-1][-1]

    def canPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        if n < 2:
            return False
        sum = 0
        max_num = 0
        for num in nums:
            sum += num
            if max_num < num:
                max_num = num
        target = sum//2

        if sum % 2 or max_num > target:
            return False
        # dp[i][j]表示0...i中和为j的情况
        dp = [True]+[False]*target

        for i in range(1, n):
            for j in range(target, nums[i]-1, -1):
                if dp[target]:
                    return True
                dp[j] = dp[j-nums[i]] | dp[j]
        return dp[-1]


# @lc code=end
if __name__ == "__main__":
    dp = [[True]+[False]*3 for _ in range(4)]
    print(dp)
